### DISCUSSION TITLE
Automap stopped working with MySQL upgrade, from 5.7 to 8.0

### DISCUSSION THREAD

**Discussion by emburse-manikamidha** [sqlalchemy/sqlalchemy#11967] [Usage Questions]
Hello

Automap worked fine with MySQL 5.7. The database at work got upgraded to 8.0 and my script seems to get into an infinite loop of some kind when control reaches the Base.prepare() stage. Has anything changed ? If it makes any difference, we are using Amazon Aurora to host our MySQL database.

My code : 

```
from sqlalchemy import create_engine
from sqlalchemy.ext.automap import automap_base

Base = automap_base()

url_object_ = mysql+pymysql://username:password@host_name:port/database_name
engine = create_engine(url_object_, future=True)
Base.prepare(autoload_with=engine, reflection_options={'views': True})
```

**Comment by CaselIT** [comment]
Hi,

As far as I know the db bersion does not change the reflection behaviour but I may be mistaken. Can you activate echo=True (or even echo='debug' to also get a log of the rows returned from the db) in the create_engine call to see if there are some odd queries being done?

It would be best if you could compare the queries in 8.0 with the ones in 5.7 to see what's different

**Reply by emburse-manikamidha** [reply]
Thanks @CaselIT I added echo=True to two versions (1 with MYSQL 5.7 and other with MySQL 8.0). Here is the difference in output:

**5.7**
2024-10-08 18:14:14,982 INFO sqlalchemy.engine.Engine SHOW CREATE TABLE `z_umsg`
2024-10-08 18:14:14,982 INFO sqlalchemy.engine.Engine [raw sql] {}
2024-10-08 18:14:14,985 INFO sqlalchemy.engine.Engine DESCRIBE `z_umsg`
2024-10-08 18:14:14,985 INFO sqlalchemy.engine.Engine [raw sql] {}
2024-10-08 18:14:15,871 INFO sqlalchemy.engine.Engine ROLLBACK

Above is the end of the trace. There are many tables in the database.

**8.0** 

There are many tables in the database. Below is just a snapshot. The lines in bold were not present when I used 5.7.
2024-10-08 18:06:23,618 INFO sqlalchemy.engine.Engine SHOW CREATE TABLE `vw_geography`
2024-10-08 18:06:23,618 INFO sqlalchemy.engine.Engine [raw sql] {}
2024-10-08 18:06:23,619 INFO sqlalchemy.engine.Engine DESCRIBE `vw_geography`
2024-10-08 18:06:23,619 INFO sqlalchemy.engine.Engine [raw sql] {}
2024-10-08 18:06:23,629 INFO sqlalchemy.engine.Engine 
                    select table_schema, table_name, column_name
                    from information_schema.columns
                    where (table_schema, table_name, lower(column_name)) in
                    ((%(table_data_1_1)s, %(table_data_1_2)s, %(table_data_1_3)s), (%(table_data_2_1)s, %(table_data_2_2)s, %(table_data_2_3)s));
                
**2024-10-08 18:06:23,630 INFO sqlalchemy.engine.Engine [generated in 0.00022s] {'table_data_1_1': 'db_name', 'table_data_1_2': 'table_name', 'table_data_1_3': 'col2', 'table_data_2_1': 'db_name', 'table_data_2_2': 'table_name', 'table_data_2_3': 'col_name'}
2024-10-08 18:06:34,580 INFO sqlalchemy.engine.Engine 
                    select table_schema, table_name, column_name
                    from information_schema.columns
                    where (table_schema, table_name, lower(column_name)) in
                    ((%(table_data_1_1)s, %(table_data_1_2)s, %(table_data_1_3)s));**

 I had to abort the Jenkins build but the lines in bold keep on printing.

**Reply by emburse-manikamidha** [reply]
@CaselIT Looks like below select statement just keeps on repeating.
```
select table_schema, table_name, column_name
from information_schema.columns
where (table_schema, table_name, lower(column_name)) in
 ((%(table_data_1_1)s, %(table_data_1_2)s, %(table_data_1_3)s), (%(table_data_2_1)s, %(table_data_2_2)s, %(table_data_2_3)s));
```

**Reply by CaselIT** [reply]
Are the parameters always the same?

**Reply by emburse-manikamidha** [reply]
The scripts accepts two arguments - customer id and env to run in. As database is different when using MySQL 5.7 and 8.0, values for customer id and env are different.

**Reply by emburse-manikamidha** [reply]
From what I can understand, above statements are printed as part of Base.prepare()

**Reply by CaselIT** [reply]
I'm not sure what script you are referring here.

You mentioned that 
> Looks like below select statement just keeps on repeating.

That statement has a series of parameters:
```sql
where (table_schema, table_name, lower(column_name)) in
 ((%(table_data_1_1)s, %(table_data_1_2)s, %(table_data_1_3)s), (%(table_data_2_1)s, %(table_data_2_2)s, %(table_data_2_3)s));
```

Are the parameter always the same? Sqlalchemy will log these in the line after. Like in your previous log above
```
2024-10-08 18:06:23,630 INFO sqlalchemy.engine.Engine [generated in 0.00022s] {'table_data_1_1': 'db_name', 'table_data_1_2': 'table_name', 'table_data_1_3': 'col2', 'table_data_2_1': 'db_name', 'table_data_2_2': 'table_name', 'table_data_2_3': 'col_name'}
```

**Reply by emburse-manikamidha** [reply]
@CaselIT I am not explicitly executing the select statements. It happens as part of Base.prepare(). These select statements get triggered when using MySQL 8.0 but not when using MySQL 5.7.

**Reply by CaselIT** [reply]
looking at the code that does that call, it seems that it enables for 8+, but it's unlikely to be a bug. Have you tried letting the code run?
How many tables are you trying to reflect?

**Reply by CaselIT** [reply]
> @CaselIT I am not explicitly executing the select statements. It happens as part of Base.prepare(). These select statements get triggered when using MySQL 8.0 but not when using MySQL 5.7.

yes I know. You mentioned that it loops. I asked to check if it's using different parameters of now. They will be logged together with the statement

**Reply by emburse-manikamidha** [reply]
> looking at the code that does that call, it seems that it enables for 8+, but it's unlikely to be a bug. Have you tried letting the code run? How many tables are you trying to reflect?
There are 614 tables.

**Reply by emburse-manikamidha** [reply]
> > @CaselIT I am not explicitly executing the select statements. It happens as part of Base.prepare(). These select statements get triggered when using MySQL 8.0 but not when using MySQL 5.7.
> 
> yes I know. You mentioned that it loops. I asked to check if it's using different parameters of now. They will be logged together with the statement

I did not understand the part about different parameters.

**Reply by emburse-manikamidha** [reply]
> > looking at the code that does that call, it seems that it enables for 8+, but it's unlikely to be a bug. Have you tried letting the code run? How many tables are you trying to reflect?
> > There are 614 tables.

I have been aborting the build as it just keeps on going. I will try to limit the tables that I need reflected and check the output.

**Reply by CaselIT** [reply]
> I did not understand the part about different parameters.

This sql has paremeters:
```sql
select table_schema, table_name, column_name
from information_schema.columns
where (table_schema, table_name, lower(column_name)) in
((%(table_data_1_1)s, %(table_data_1_2)s, %(table_data_1_3)s), (%(table_data_2_1)s, %(table_data_2_2)s, %(table_data_2_3)s));
```
so it can be invoked many times with different paramters. What I'm asking if the paramters keep chaning or not.

In any case if you can restrict the number of tables that are reflected it should speed up the reflection

**Reply by emburse-manikamidha** [reply]
Yes, the params change for each select statement. I could get the build to run successfully after limiting the number of tables. It still takes time to run but it runs.

**Reply by CaselIT** [reply]
understood. so it seems to be just slow.

did the old mysql run significantly faster?

**Reply by emburse-manikamidha** [reply]
Yes, a lot faster. With old mysql 5.7, Automap scanned 600+ tables in barely noticeable time. To scan 12 tables using Mysql 8.0, iAutomap, took 4 minutes.

**Reply by CaselIT** [reply]
that's very strange, since we don't have any other report about this. it's not normal that it took 4 minutes
is the db schema you are trying to reflect public by any chance?

**Reply by emburse-manikamidha** [reply]
I would imagine the database schema to be private to my company. The same database worked fine with MySQL 5.7 but not with MySQL 8.0. My concern is , why select queries are emitted as part of Base.prepare(). Just a thought that slowdown could be because of that. I did not see these select queries when using Automap with MySQL 5.7

**Reply by CaselIT** [reply]
Prepare needs to reflect the database schema. 

I'm not sure how you suppose it could do that without using selects to the information tables of the db.

There is a different logic on mysql 8 that adds those select.
May the wort a try to disable it just to check it it's the issue.
Can you try doing and check if the are any changes?
```py
engine = create_engine(url_object_, future=True)
engine.connect().close()
engine.dialect._needs_correct_for_88718_96365 = False
Base.prepare(autoload_with=engine, reflection_options={'views': True})
```

The above script should disable the extra select used to correct issues on mysql 8+

**Reply by emburse-manikamidha** [reply]
@CaselIT This worked like a charm.  In the output, I do not see select statements anymore.Can you please share what exactly happened with these two lines : 
```
engine.connect().close()
engine.dialect._needs_correct_for_88718_96365 = False
```

**Reply by emburse-manikamidha** [reply]
I just re-read your latest comment. You said that Prepare uses Select statements to reflect the Database. I have disabled the select statements. How is the Database getting reflecting  without the Select statements ?

**Reply by CaselIT** [reply]
`engine.connect().close()` is just to initialize the dialect that happens on the first connection of the engine. (otherwise the attribute would be overwritten)
`engine.dialect._needs_correct_for_88718_96365 = False` disables the additional steps in mysql 8.0 that seems to fix an issue in that version but it seems to have performance issues here.

Just to understand the size of the db to try to reproduce, could you run the following queries?
```sql
select count(1)
from information_schema.columns
```
```sql
select count(1)
from (select distinct table_schema, table_name from information_schema.columns) as x
```

> I just re-read your latest comment. You said that Prepare uses Select statements to reflect the Database. I have disabled the select statements. How is the Database getting reflecting without the Select statements ?

mysql reflection is a mix of select and show table (that's basically a select with a different syntax)

**Reply by CaselIT** [reply]
can you also report the exact version of mysql 8 in use?

**Reply by emburse-manikamidha** [reply]
> can you also report the exact version of mysql 8 in use?

8.0.36

**Reply by emburse-manikamidha** [reply]
```
select count(1)
from information_schema.columns
```
3091788 

```
select count(1)
from (select distinct table_schema, table_name from information_schema.columns) as x
```
58316

**Reply by CaselIT** [reply]
thanks!

can you also run
```sql
select count(1)
from information_Schema.REFERENTIAL_CONSTRAINTS 
```

so that we can create a better example

**Reply by emburse-manikamidha** [reply]
> thanks!
> 
> can you also run
> 
> ```sql
> select count(1)
> from information_Schema.REFERENTIAL_CONSTRAINTS 
> ```
> 
> so that we can create a better example

1033

**Reply by CaselIT** [reply]
the issue that tracks the slow queries is #11975

**Comment by CaselIT** [comment]
@zzzeek do you remember why `_needs_correct_for_88718_96365 ` was added in mysql? Is it still needed??

It enables this: https://github.com/sqlalchemy/sqlalchemy/blob/74a8e2ced922183d6ad072eced904cb989113fa2/lib/sqlalchemy/dialects/mysql/base.py#L3042-L3127

it seems to cause large performance issues at least on the posted database (12 tables are reflected in 4 minutes).

it was originally added here https://github.com/sqlalchemy/sqlalchemy/commit/56fb68ca8620a211ca29b3d47d649dfa332d354a

What's strange is that in 6 years it's the first time we have a report like this

**Reply by zzzeek** [reply]
it works around a MySQL bug that made it impossible for us to reflect foreign key information

im not sure that it has a negative impact on a DB that does not need the fix

do we have an MCVE here?

**Reply by emburse-manikamidha** [reply]
Just to correct myself, 32 tables were reflected in 3 minutes and 35 secs using MySQL 8.0.

**Reply by CaselIT** [reply]
> do we have an MCVE here?

no but we can probably create one using there information https://github.com/sqlalchemy/sqlalchemy/discussions/11967#discussioncomment-10893458

I think it's just slow because the table is big

**Reply by CaselIT** [reply]
likely using a simpler condition solves this case

**Reply by CaselIT** [reply]
created #11975 for the performance issue

**Reply by CaselIT** [reply]
@emburse-manikamidha the slowdown should be fixed on the main and 2.0 branch and should be released in a few days. If you could later try it it would be great! (remembering to remove the `engine.dialect._needs_correct_for_88718_96365 = False` workaround)

Thanks for the help with the provided queries!

**Reply by emburse-manikamidha** [reply]
Thanks @CaselIT . MySQL 8.0.36 was reverted on my work database due to many issues and less time to resolve them. I will wait for the Database to be  upgraded to test.

**Reply by CaselIT** [reply]
ok. In any case I've tried to create a large db with a number of items similar to yours and I could clearly see that the original query was taking more than 10s. The new one completed in something like 10ms, so that was definitely something that needed to be improved

**Reply by emburse-manikamidha** [reply]
@CaselIT Sounds great.

**Reply by emburse-manikamidha** [reply]
@CaselIT I got a chance to test the update today as the Database got upgraded to 8.0.36. The queries are slow if I don't use : 
```
engine.connect().close()
engine.dialect._needs_correct_for_88718_96365 = False
```

**Reply by CaselIT** [reply]
Thanks for testing. Is this with 2.0.36? If so I don't know if there is anything more that can be done here, since the query that's currently being run should hit the index.

Overall that option adds an additional query per table, so it's expect to take more time, but it shouldn't be minutes like before to reflect only a few tables.

**Reply by emburse-manikamidha** [reply]
> Thanks for testing. Is this with 2.0.36? If so I don't know if there is anything more that can be done here, since the query that's currently being run should hit the index.
> 
> Overall that option adds an additional query per table, so it's expect to take more time, but it shouldn't be minutes like before to reflect only a few tables.

@CaselIT Yes, 8.0.36. This issue is not a show stopper for me. Adding `engine.dialect._needs_correct_for_88718_96365 = False` is good enough for me. I posted here just to share the test results. Thanks again for looking into this.

**Reply by CaselIT** [reply]
shame that's still not performant enough, but I would assume that dbs with 3.1 million columns are not very common.

thanks for testing

**Comment by EncoderPtd** [comment]
python=3.12.11
SQLAlchemy=2.0.41

MySql:
print(self.server_version_info)
(8, 0, 11, 5, 1)

```py
# not work, because it will be covered in metadata.reflect method
engine.dialect._needs_correct_for_88718_96365 = False

# this worked... 😂
engine.dialect._correct_for_mysql_bugs_88718_96365 = lambda a, b: None
```
